database2/ffldb/ff/write.go

```diff
@@ -0,0 +1,166 @@
+package ff
+
+import (
+	"github.com/pkg/errors"
+	"hash/crc32"
+	"os"
+	"syscall"
+)
+
+// write appends the specified data bytes to the store's write cursor location
+// and increments it accordingly. When the data would exceed the max file size
+// for the current flat file, this function will close the current file, create
+// the next file, update the write cursor, and write the data to the new file.
+//
+// The write cursor will also be advanced the number of bytes actually written
+// in the event of failure.
+//
+// Format: <data length><data><checksum>
+func (s *flatFileStore) write(data []byte) (*flatFileLocation, error) {
+	if s.isClosed {
+		return nil, errors.Errorf("cannot write to a closed store %s",
+			s.storeName)
+	}
+
+	// Compute how many bytes will be written.
+	// 4 bytes for data length + length of the data + 4 bytes for checksum.
+	dataLength := uint32(len(data))
+	fullLength := dataLength + 8
+
+	// Move to the next file if adding the new data would exceed the max
+	// allowed size for the current flat file. Also detect overflow because
+	// even though it isn't possible currently, numbers/ might change in
+	// the future to make it possible.
+	//
+	// NOTE: The writeCursor.offset field isn't protected by the mutex
+	// since it's only read/changed during this function which can only be
+	// called during a write transaction, of which there can be only one at
+	// a time.
+	cursor := s.writeCursor
+	finalOffset := cursor.currentOffset + fullLength
+	if finalOffset < cursor.currentOffset || finalOffset > maxFileSize {
+		// This is done under the write cursor lock since the curFileNum
+		// field is accessed elsewhere by readers.
+		//
+		// Close the current write file to force a read-only reopen
+		// with LRU tracking. The close is done under the write lock
+		// for the file to prevent it from being closed out from under
+		// any readers currently reading from it.
+		cursor.Lock()
+		cursor.currentFile.Lock()
+		if cursor.currentFile.file != nil {
+			_ = cursor.currentFile.file.Close()
+			cursor.currentFile.file = nil
+		}
+		cursor.currentFile.Unlock()
```

Comment:
I see that we have a ton of Unlocks without defer, so this is last time I write about it, simply go over the code and make sure that unlocks are always deferred.

Reply:
Created a ticket.